_posts/2020-06-10-connect-kafka-spring-kafka.md

```diff
@@ -0,0 +1,575 @@
+    }
+
+    @Bean
+    public KafkaTemplate<String, User> userKafkaTemplate() {
+        return new KafkaTemplate<>(userProducerFactory());
+    }
+}
+```
+
+```java
+@Configuration
+class KafkaConsumerConfig {
+    ...
+    public ConsumerFactory<String, User> userConsumerFactory() {
+        Map<String, Object> props = new HashMap<>();
+        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, "reflectoring-user");
+
+        return new DefaultKafkaConsumerFactory<>(props, new StringDeserializer(), new JsonDeserializer<>(User.class));
+    }
+
+    @Bean
+    public ConcurrentKafkaListenerContainerFactory<String, User> userKafkaListenerContainerFactory() {
+        ConcurrentKafkaListenerContainerFactory<String, User> factory = new ConcurrentKafkaListenerContainerFactory<>();
+        factory.setConsumerFactory(userConsumerFactory());
+        return factory;
+    }
+    ...
+}
+```
+
+Spring Kafka provides `JsonSerializer` and `JsonDeserializer` implementations that are based on the Jackson JSON object mapper. It allows us to convert any Java Object to `bytes[]`.
+
+In the above example, we are creating one more `ConcurrentKafkaListenerContainerFactory` for JsonSerialization. In this, we have configured `JsonSerializer.class` as our value serializer in Producer config and `JsonDeserializer<>(User.class)` as out value deserializer in consumer config. For this, we are creating separate Kafka listener container `userKafkaListenerContainerFactory`. If we have multiple Java object types to be serialized/deserialized, we have to create a listener container for each type as shown above.
+
+Now that we have configured our serializer & deserializer, we can send a User object using `KafkaTemplate`. We'll be defining our sender implementations throughout the article in `KafkaSenderExample`:
+
+```java
+@Component
+class KafkaSenderExample {
+  ...
+
+  @Autowired
+  private KafkaTemplate<String, User> userKafkaTemplate;
+
+  void sendCustomMessage(User user, String topicName) {
+    userKafkaTemplate.send(topicName, user);
+  }
+  ...
+}
+```
+
+As shown above, we have to create the `KafkaTemplate` instance for the specific type and send user records to designated topics by passing them as parameters.
+
+```java
+@Component
+class KafkaListenersExample {
+
+  @KafkaListener(topics = "reflectoring-user", groupId="reflectoring-user", containerFactory="userKafkaListenerContainerFactory")
+  void listener(User user) {
+    LOG.info("CustomUserListener [{}]", user);
+  }
+}
+```
+
+Since we have multiple listener containers, we are specifying which container factory (`containerFactory="userKafkaListenerContainerFactory"`) to be used while consuming records.
+
+If we don't specify this attribute it defaults to `kafkaListernerContainerFactory` which uses `StringSerializer` and `StringDeserializer` in our case.
+
+### Using Spring's Message Conversion
```

Comment:
I'm sorry, but from the text in this section, I don't understand how Spring message conversion works with KafkaTemplate.  What is it good for?

I have removed this section as the article is long enough as it is.

Reply:
It's ok. We have Json Serailization approach which is enough I guess.